File: homework8/static/components/noteshandler/noteshandler.js

// The following code is written and used almost exactly the same from Matan Broner's TA session
// Please grade accordingly

(function () {
    var noteshandler = {
        props: ["get_all_notes", "create_new_note", "edit_note_key", "delete_note", "note_set_color"],
        data: {},
        methods: {},
    };

    const colors_classes = ["blue", "purple", "red", "green", "yellow", "white"];

    // handle the data in the note
    noteshandler.data = function () {
        var data = {
            urls: {
                get_all: this.get_all_notes,
                create_new: this.create_new_note,
                delete_note: this.delete_note,
                edit_key: this.edit_note_key,
                set_color: this.note_set_color,
            },
            notes: [],
            colors_classes,
            flag: false,
        };
        noteshandler.methods.load.call(data);
        return data;
    };


    // API CALLS

    // load
    noteshandler.methods.load = function () {
        let self = this;
        axios.get(self.urls.get_all).then((res) => {
            let notes = res.data.notes;
            notes.forEach((note) => (note.edit = false));
            self.notes = notes;
            reindex(self.notes);
            console.log("load called\n");
        });
    };

    // create note
    noteshandler.methods.create = function () {
        let self = this;
        axios.get(self.urls.create_new).then((res) => {
            let note = res.data.note;
            note.edit = false;
            self.notes.unshift(note);
            reindex(self.notes);
            console.log("Created note\n", note);
        });

    };

    //UTILS

    noteshandler.methods.remove = function (idx) {
        let self = this;
        let id = self.notes[idx].id;
        axios.post(self.urls.delete_note, { id }).then((res) => {
            this.notes.splice(idx, 1);
            reindex(self.notes);
            console.log("remove() called\n");
        });
    };

    noteshandler.methods.edit_key = function (idx, key, val) {
        let self = this;
        axios.post(self.urls.edit_key, {
            id: self.notes[idx].id,
            key,
            val,
        })
            .then((res) => {
                self.notes[idx] = {
                    ...res.data.note,
                    _idx: idx,
                    edit: true,
                };
                this.refresh();
            });
    };



    // reindex
    function reindex(items) {
        for (let i = 0; i < items.length; i++) {
            items[i]._idx = i;
            items[i].edit = false;
        }

    }

    function sort_date(items) {
        return items.sort(function (a, b) {
            return new Date(b.last_modified) - new Date(a.last_modified);
        });
    }


    // Compent meeeethodsss

    // format_note_text
    noteshandler.methods.format_note_text = function (content, idx) {
        if (this.notes[idx].is_list) {
            return content
                .split("\n") // create array of string broken at each newline
                .map((item) => {
                    // add a bullet + space to each array item
                    let format = "• " + item;
                    return format;
                })
                .join("\n"); // join the strings array as one string with newline
        } else return content;
    };


    // handle_note_change
    noteshandler.methods.handle_note_change = function (event, idx) {
        // Alternatively, use a replace() while indexOf character is found
        let clean = event.target.value.replace(/• /g, "").replace(/•/g, ""); // replace bullets with and without space
        this.notes[idx].content = clean;
        this.edit_key(idx, "content", clean);
    };


    noteshandler.methods.complete_edit = function (idx) {
        this.toggle_edit(false, idx);
        let notes = this.sort_notes();
        reindex(notes);
        this.notes = notes;
    }

    // toggle_edit
    noteshandler.methods.toggle_edit = function (bool, idx) {
        if (bool === true && this.notes.findIndex((note) => note.edit) !== -1) {
            return;
        }
        this.$set(this.notes, idx, { ...this.notes[idx], edit: bool });
    };

    noteshandler.methods.color_class = function (index) {
        return "pastel-" + this.colors_classes[index];
    };

    noteshandler.methods.sort_notes = function () {
        let starred = this.notes.filter((note) => note.has_star === true);
        let not_starred = this.notes.filter((note) => note.has_star === false);

        return [...sort_date(starred), ...sort_date(note_starred)];
    };

    noteshandler.methods.set_color = function (note, id, color) {
        let self = this;
        axios.post(self.urls.set_color, {
            id: id,
            color: color,
            note: note,
        })
        .then(() => {
            note.color = color
        });
        this.refresh();
        console.log("set_color() called\n");
    }



    // force refresh the component state
    noteshandler.methods.refresh = function (idx) {
        this.$forceUpdate();
    };


    // register the vue component
    utils.register_vue_component(
        "notes",
        "components/noteshandler/noteshandler.html",
        function (template) {
            noteshandler.template = template.data;
            return noteshandler;
        }
    );

})();

File: homework8/static/components/noteshandler/noteshandler.py
# The following code is written and used almost exactly the same from Matan Broner's TA session
# Please grade accordingly


from py4web import action, request, URL
from yatl.helpers import XML
from py4web.utils.url_signer import URLSigner
from py4web.core import Fixture



# NotesHandler class (component)
class NotesHandler(Fixture):
    # vue tag we are going to be using with 2 routes
    NOTESHANDLER = """<notes
    get_all_notes={get_all_notes}
    create_new_note={create_new_note}
    delete_note={delete_note}
    edit_note_key={edit_note_key}
    note_set_color={note_set_color}
    
    ></notes>"""

    # constructor
    def __init__(self, url, session, signer=None, db=None, auth=None):
        self.db = db
        self.url_base = url
        self.signer = signer or URLSigner(session)

        self.define_urls()

        self.__prerequisites__ = [session]
        args = list(filter(None, [session, db, auth, self.signer.verify()]))

        self.define_route(args, self.create_new_note_url, self.create_new_note, ["GET"])
        self.define_route(args, self.delete_note_url, self.delete_note, ["POST"])
        self.define_route(args, self.get_all_notes_url, self.get_all_notes, ["GET"])
        self.define_route(args, self.edit_note_key_url, self.edit_note_key, ["POST"])
        self.define_route(args, self.note_set_color_url, self.set_color, ["POST"])


    def __call__(self):
        return XML(NotesHandler.NOTESHANDLER.format(
            get_all_notes=URL(self.get_all_notes_url, signer=self.signer),
            create_new_note=URL(self.create_new_note_url, signer=self.signer),
            delete_note=URL(self.delete_note_url, signer=self.signer),
            edit_note_key=URL(self.edit_note_key_url, signer=self.signer),
            note_set_color=URL(self.note_set_color_url, signer=self.signer)
        ))

    # this was in the last homework
    def define_route(self, args, url, class_method, method):
        func = action.uses(*args)(class_method)
        action(url, method=method)(func)

    def define_urls(self):
        self.get_all_notes_url = self.url_base
        self.create_new_note_url = self.url_base + "/new"
        self.delete_note_url = self.url_base + "/delete"
        self.edit_note_key_url = self.url_base + "/edit"
        self.note_set_color_url = self.url_base + "/edit"

    # takes the db
    # self is set in the constructor
    def get_all_notes(self):
        notes = self.db(self.db.notes).select().as_list()
        return dict(notes=notes)

    # creates a new note with no field set and returns the new note
    # as a dictionary at a specific index (id)
    def create_new_note(self):
        id = self.db.notes.insert()
        return dict(note=self.db.notes[id].as_dict())

    def set_star(self):
        # PUT STUFF HERE
        return

    def delete_note(self):
        id = request.json["id"]
        self.db(self.db.notes.id == id).delete()
        return "ok" # shouldn't get used

    def edit_note_key(self):
        id = request.json["id"]
        key = request.json["key"]
        val = request.json["val"]
        update_dict = {
            key: val
        }

        self.db(self.db.notes.id == id).update(**update_dict)
        return dict(note=self.db.notes[id].as_dict())

    def set_color(self):
        id = request.json["id"]
        color = request.json["color"]
        self.db(self.db.notes.id == id).update(color = color)
        return "ok"


# if __name__ == "__main__":
#     handler = NotesHandler('notes', session, db=db)
#     pass